generate_route.py

```diff
@@ -81,36 +81,112 @@ def __init__(self, start_location, end_location):
+    """
     locations,self.distances,self.durations,self.polyline = request_directions(self.start_location.get_lat_lon_tuple(),self.end_location.get_lat_lon_tuple())
     route = []
     for location in locations:
       latitude = location[0]
       longitude = location[1]
       route.append(Location(latitude,longitude))
     self.route = route
+    self.add_altitudes()
   
   def add_altitudes(self):
+    """Add altitudes to each point in self.route using Maps Elevation API.
+
+    sets elevation data for each point, and then adds xyz conversion from lla to each point
+    """
     locations = []
     for location in self.route:
       locations.append(location.get_lat_lon_tuple())
     elevations = request_elevations(locations)
     for location,elevation in zip(self.route,elevations):
       location.altitude = elevation
+      location.add_XYZ()
+
+  def write_route(self,file_name):
+    """write route into csv with each line as x,y,z.
+
+    """
+    write_array = []
+    for location in self.route:
+      write_array.append(location.get_xyz_tuple())
+    write_to_csv(file_name,write_array)
+
+class TimedRoute(Route):
+  """An object for a route that has a desired speed and point frequency
+
+  Attributes:
+      start_location: a Location object for the start of the route
+      end_location: a Location object for the end of the route
+      speed: how fast the person moves through the route
+      frequency: how many points per second the timed route should have (Hz)
+      route: a list of Location objects for each point on the route
+      distances: a list of distances between each pair of consecutive locations
+      durations: a list of durations between each pair of consecutive locations
+      polyline: an encoded format for the route given by the Maps API for ease of drawing
+  """
+  def __init__(self, start_location, end_location,speed,frequency):
+    Route.__init__(self,start_location,end_location)
+    self.speed = speed
+    self.frequency = frequency
+
+  def create_route(self):
+    Route.create_route(self)
+    self.upsample_route()
   
   def upsample_route(self):
-    pass
+    """Upsample the TimedRoute such that the number of points aligns with the desired
+    speed and frequency.
+
+    for each consecutive set of points, the change in lat,lon,alt is divided by
+    split amongst the number of new points that need to be created so that there
+    is roughly an equal distance (1/points_per_meter) between each of the points in the 
+    upsampled route.
+    """
+    points_per_meter = self.frequency/self.speed
+    new_route = []
+
+    # fill first 10 cycles with first location
+    for i in range(10):
+      new_route.append(self.route[0])
+
+    for i in range(len(self.distances)):
+      distance = self.distances[i]
+      start_point = self.route[i]
+      end_point = self.route[i+1]
+
+      points_needed = int(distance*points_per_meter)-1
```

Comment:
current behavior is still just upsampling (for a 30 minute walk those values produce ~1100 points). But if points_needed dropped below 1, then it would originally let it "downsample" and get rid of the original point, but I think I'll change it so that at the very least the original points are kept, so essentially not allowing for downsampling